xyz

```diff
@@ -117,62 +122,69 @@ inc() {
   printf %s "$version"
 }
 
-branch=master
-edit=false
-increment=patch
-message_template='Version X.Y.Z'
-prerelease_label=rc
-publish_command='npm publish'
-repo=origin
+declare -A options=(
+  [branch]=master
+  [dry-run]=false
+  [edit]=false
+  [increment]=patch
+  [message]='Version X.Y.Z'
+  [prerelease-label]=rc
+  [publish-command]='npm publish'
+  [repo]=origin
+  [tag]=vX.Y.Z
+)
 declare -a scripts
-tag_template=vX.Y.Z
-dry_run=false
 
 while (( $# > 0 )) ; do
-  option="$1"
-  shift
+  option="$1" ; shift
+
+  case $option in
+    -b) option=--branch ;;
+    -e) option=--edit ;;
+    -h) option=--help ;;
+    -i) option=--increment ;;
+    -m) option=--message ;;
+    -r) option=--repo ;;
+    -t) option=--tag ;;
+    -v) option=--version ;;
+  esac
 
-  case "$option" in
-    -h|--help)
-      echo "$usage"
-      exit
-      ;;
-    -v|--version)
-      node -p 'require("xyz/package.json").version'
-      exit
-      ;;
-    -b|--branch)            branch="$1"           ; shift ;;
-    -e|--edit)              edit=true             ;;
-    -i|--increment)         increment="$1"        ; shift ;;
-    -m|--message)           message_template="$1" ; shift ;;
-       --prerelease-label)  prerelease_label="$1" ; shift ;;
-       --publish-command)   publish_command="$1"  ; shift ;;
-    -r|--repo)              repo="$1"             ; shift ;;
-    -s|--script)            scripts+=("$1")       ; shift ;;
-    -t|--tag)               tag_template="$1"     ; shift ;;
-       --dry-run)           dry_run=true          ;;
+  case $option in
+    --help)
+      echo "$usage" ; exit ;;
+    --version)
+      node -p 'require("xyz/package.json").version' ; exit ;;
+    --script)
+      scripts+=("$1") ; shift ;;
+    --dry-run|--edit)
+      options[${option:2}]=true ;;
     *)
-      echo "Unrecognized option $option" >&2
-      exit 1
+      if [[ $option =~ ^--(.+)$ && -n ${options[${BASH_REMATCH[1]}]+x} ]] ; then
+        options[${BASH_REMATCH[1]}]="$1" ; shift
+      else
+        abort "Unrecognized option $option"
+      fi
```

Comment:
`-n ${options[${BASH_REMATCH[1]}]+x}` requires an explanation.

[`BASH_REMATCH`][1]:

> An array variable whose members are assigned by the ‘`=~`’ binary operator to the `[[` conditional command (see [Conditional Constructs][2]). The element with index 0 is the portion of the string matching the entire regular expression. The element with index *n* is the portion of the string matching the *n*th parenthesized subexpression. This variable is read-only.

So, if option is `--branch`, `${BASH_REMATCH[1]}` will be `branch`.

We then wish to determine whether `branch` is a key of `options`. The naive approach is to test the result of the lookup with `-n`, but this assumes the null string will never be used as a value. The thing we actually want to determine is whether the key *exists*. For this we can use parameter expansion:

```bash
declare -A arr=([foo]=FOO [bar]='')

echo "foo: ${arr[foo]+x}" # foo: x
echo "bar: ${arr[bar]+x}" # bar: x
echo "baz: ${arr[baz]+x}" # baz:
```

So, `${options[${BASH_REMATCH[1]}]+x}` will evaluate to `x` if the specified option exists; the null string otherwise. We then use `-n` to differentiate these two cases.


[1]: https://www.gnu.org/software/bash/manual/html_node/Bash-Variables.html#index-BASH_005fREMATCH
[2]: https://www.gnu.org/software/bash/manual/html_node/Conditional-Constructs.html#Conditional-Constructs